File: 2799-count-complete-subarrays-in-an-array/2799-count-complete-subarrays-in-an-array.py
class Solution:
    def countCompleteSubarrays(self, nums: List[int]) -> int:
        mp, l, res, size = {}, 0, 0, len(set(nums))
        for r, c in enumerate(nums):
            mp[c] = mp.get(c, 0) + 1
            if mp[c] == 1:
                size -= 1
            while size == 0:
                mp[nums[l]] -= 1
                if mp[nums[l]] == 0:
                    size += 1
                l += 1
            res += l
        return res

File: 2799-count-complete-subarrays-in-an-array/2799-count-complete-subarrays-in-an-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var countCompleteSubarrays = function(nums) {
    let required = (new Set(nums)).size
    let res = 0
    let l = 0
    const map = {}
    for (let r=0; r < nums.length; r++) {
        map[nums[r]] = map[nums[r]] + 1 || 1
        if (map[nums[r]] === 1) required--
        while (required === 0) {
            map[nums[l]]--
            if (map[nums[l]] === 0) required++
            l++
        }
        res += l
    }
    return res
};